application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java

```diff
@@ -0,0 +1,118 @@
+import uk.nhs.hee.tis.revalidation.repository.RecommendationRepository;
+import uk.nhs.hee.tis.revalidation.repository.SnapshotRepository;
+
+@Slf4j
+@Service
+public class RecommendationTisStatusUpdateService {
+
+  @Autowired
+  private RecommendationRepository recommendationRepository;
+  @Autowired
+  private DoctorsForDBRepository doctorsForDBRepository;
+  @Autowired
+  private SnapshotRepository snapshotRepository;
+  @Autowired
+  private RecommendationService recommendationService;
+  @Autowired
+  private SnapshotService snapshotService;
+
+  public void updateRecommendationAndTisStatus(
+      final RecommendationStatusCheckDto recommendationStatusCheckDto) {
+    RecommendationGmcOutcome outcome = recommendationStatusCheckDto.getOutcome();
+    setRecommendationStatusAndUpdateRepositories(recommendationStatusCheckDto.getRecommendationId(),
+        outcome, recommendationStatusCheckDto.getGmcRecommendationId());
+
+    String gmcReferenceNumber = recommendationStatusCheckDto.getGmcReferenceNumber();
+    Optional<DoctorsForDB> optionalDoctorsForDB = doctorsForDBRepository
+        .findById(gmcReferenceNumber);
+
+    //update the tis status in doctorsfordb to "complete"
+    if (optionalDoctorsForDB.isPresent() && (APPROVED.equals(outcome) || REJECTED
+        .equals(outcome))) {
+      final var doctorsForDB = optionalDoctorsForDB.get();
+      doctorsForDB.setDoctorStatus(
+          recommendationService.getRecommendationStatusForTrainee(gmcReferenceNumber));
+      doctorsForDBRepository.save(doctorsForDB);
+    }
+  }
+
+  //if gmc outcome is approved/rejected, update the relevant recommendation to approved/rejected,
+  //tis status to complete and also update the snapshot repository
+  private void setRecommendationStatusAndUpdateRepositories(final String recommendationId,
+      final RecommendationGmcOutcome recommendationGmcOutcome, final String gmcRecommendationId) {
+
+    final var optionalRecommendation = recommendationRepository.findById(recommendationId);
+
+    if (optionalRecommendation.isPresent()) {
+      Recommendation recommendation = optionalRecommendation.get();
+      if (APPROVED.equals(recommendationGmcOutcome) || REJECTED.equals(recommendationGmcOutcome)) {
+        recommendation.setOutcome(recommendationGmcOutcome);
+        recommendation.setRecommendationStatus(COMPLETED);
+        recommendationRepository.save(recommendation);
+
+        // Find the snapshot from snapshot repository and check the gmcRecommendationId.
+        // If it exists then don't save snapshot else save it
+        if (!doesSnapshotRecommendationExist(recommendation.getGmcNumber(),
+            gmcRecommendationId)) {
+          recommendation.setGmcRevalidationId(gmcRecommendationId);
+          snapshotService.saveRecommendationToSnapshot(recommendation);
+        }
+      }
+    }
+  }
+
+  private boolean doesSnapshotRecommendationExist(String gmcReferenceNumber,
```

Comment:
Just wondering, Do I need to make this method public in order to test? Or how do we test private method in mockito?

Reply:
IMHO, contents in private methods can be covered in the tests of the public method that invoke it (updateRecommendationAndTisStatus()). Just write test cases to cover all the conditions (both happy and unhappy paths) can do.

Reply:
In this [sonar coverage page ](https://sonarcloud.io/component_measures?id=Health-Education-England_tis-revalidation-recommendation&metric=new_coverage&pullRequest=315&selected=Health-Education-England_tis-revalidation-recommendation%3Aapplication%2Fsrc%2Fmain%2Fjava%2Fuk%2Fnhs%2Fhee%2Ftis%2Frevalidation%2Fservice%2FRecommendationTisStatusUpdateService.java&view=list), 
the red bar in front of the code indicates which lines are not covered. Mouse over the red bar, it would tell you the condition is partially covered. You may write different test cases to cover all paths.

Reply:
I have added another test case with REJECTED

Reply:
`REJECTED` doesn't cover this alternate path.